[SPARK-18471][MLLIB] In LBFGS, avoid sending huge vectors of 0

AnthonyTruchet opened the pull request:
## What changes were proposed in this pull request?

CostFun used to send a dense vector of zeroes as a closure in a
treeAggregate call. To avoid that, we replace treeAggregate by
mapPartition + treeReduce, creating a zero vector inside the mapPartition
block in-place.

## How was this patch tested?

Unit test for module mllib run locally for correctness.

As for performance we run an heavy optimization on our production data (50 iterations on 128 MB weight vectors) and have seen significant decrease in terms both of runtime and container being killed by lack of off-heap memory.

AnthonyTruchet commented:
See #15963 for first feed back on those changes.

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented:
OK, this is the fourth pull request though (not counting a not-quite-related 5th). You don't need to open a new PR to push more changes and it adds to the difficulty in reviewing. This still doesn't incorporate suggestions from the last review.

srowen commented:
Following https://github.com/apache/spark/pull/16038 I suggest this proceed by making the zero value a sparse vector, and then making it dense in the seqOp immediately.

MLnick commented:
This is all a bit confusing - can we highlight which PR is actually to be reviewed?

srowen commented:
This should be the main PR @MLnick

MLnick commented:
Right ok. So I think the approach of making the zero vector sparse then calling `toDense` in `seqOp` as @srowen suggested makes most sense.

Currently the gradient vector *must* be dense in MLlib since both `axpy` and the logic for multinomial logreg requires it. So the thing that is initially serialized with the task should be tiny, and the call `toDense` for the first instance in each partition will essentially generate the dense zero vector. Thereafter it should be a no-op as the vector will be dense and `toDense` will just be a ref to the values array.

Can we see if this works:
```scala
      val zeroVector = Vectors.sparse(n, Seq())
      val (gradientSum, lossSum) = data.treeAggregate((zeroVector, 0.0))(
          seqOp = (c, v) => (c, v) match { case ((grad, loss), (label, features)) =>
            val denseGrad = grad.toDense
            val l = localGradient.compute(
              features, label, bcW.value, denseGrad)
            (denseGrad, loss + l)
          },
          combOp = (c1, c2) => (c1, c2) match { case ((grad1, loss1), (grad2, loss2)) =>
            axpy(1.0, grad2, grad1)
            (grad1, loss1 + loss2)
          })
```

MLnick commented:
What worries me more actually is that the initial vector when sent in the closure should be compressed. So why is this issue occurring? Is it a problem with serialization / compression? OR even after compression it is still too large? Would be good to understand that.

AnthonyTruchet commented:
Hello @MLnick,

Sorry that my push of a new version just crossed with your suggestion.

I'll test you suggestion. Thanks to for having written it, I didn't get that Sean's hint was this.

AnthonyTruchet commented:
@MLnick I have to add a transtyping denseGrad in the expression returned by seqOp from DenseVector to Vector. The only way I could find to do it with the API are: `Vectors.fromBreeze(denseGrad.asBreeze)` or `Vectors.dense(denseGrad.values)`: I'm a biut concerned with the potential recopy of a 128MB piece of memory...

Commit:
[SPARK-18471][MLLIB] In LBFGS, avoid sending huge vectors of 0

CostFun used to send a dense vector of zeroes as a closure in a
treeAggregate call. To avoid that, we ensure the zero is sent as a
SparseVector but that it is converted into a DensseVector immediately in
seqOp, thus effectively generating a (potentially huge) vector of zero on
the executors only.

MLnick commented:
ok to test

SparkQA commented:
**[Test build #69463 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69463/consoleFull)** for PR 16037 at commit [`d7ebc7d`](https://github.com/apache/spark/commit/d7ebc7df63b89d7ba7cd7e3a688089749c393082).

SparkQA commented:
**[Test build #69463 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69463/consoleFull)** for PR 16037 at commit [`d7ebc7d`](https://github.com/apache/spark/commit/d7ebc7df63b89d7ba7cd7e3a688089749c393082).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69463/
Test FAILed.

AnthonyTruchet commented:
Scala style checks fixed, and spurious cast removed.

SparkQA commented:
**[Test build #69466 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69466/consoleFull)** for PR 16037 at commit [`0ce8c64`](https://github.com/apache/spark/commit/0ce8c644e3cbed3cf600fcb30526e46a8054e498).

srowen reviewed (approved):
Yes this LGTM. Let's make sure it passes tests. CC maybe @dbtsai FYI

SparkQA commented:
**[Test build #69466 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69466/consoleFull)** for PR 16037 at commit [`0ce8c64`](https://github.com/apache/spark/commit/0ce8c644e3cbed3cf600fcb30526e46a8054e498).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69466/
Test FAILed.

AnthonyTruchet commented:
This seems very related to the change. I know nothing wrt the way PySpark interfaces Python and Scala and I'm surprised that changing an internal f the Scala lib causes this ? Ready to learn and help though :-)

```
File "/home/jenkins/workspace/SparkPullRequestBuilder/python/pyspark/mllib/classification.py", line 155, in __main__.LogisticRegressionModel
Failed example:
    mcm = LogisticRegressionWithLBFGS.train(data, iterations=10, numClasses=3)
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib64/python2.6/doctest.py", line 1253, in __run
        compileflags, 1) in test.globs
      File "<doctest __main__.LogisticRegressionModel[23]>", line 1, in <module>
        mcm = LogisticRegressionWithLBFGS.train(data, iterations=10, numClasses=3)
      [...]
    Py4JJavaError: An error occurred while calling o173.trainLogisticRegressionModelWithLBFGS.
     [...]
Caused by: java.lang.IllegalArgumentException: axpy only supports adding to a dense vector but got type class org.apache.spark.mllib.linalg.SparseVector.
    	at org.apache.spark.mllib.linalg.BLAS$.axpy(BLAS.scala:58)
    	at org.apache.spark.mllib.optimization.LBFGS$CostFun$$anonfun$2.apply(LBFGS.scala:257)
    	at org.apache.spark.mllib.optimization.LBFGS$CostFun$$anonfun$2.apply(LBFGS.scala:255)
```

Commit:
Fix error triggered in PySpark

MLnick commented on mllib/src/main/scala/org/apache/spark/mllib/optimization/LBFGS.scala:
```diff
@@ -241,16 +241,27 @@ object LBFGS extends Logging {
+          val denseGrad1 = grad1.toDense
+          val denseGrad2 = grad2.toDense
+          axpy(1.0, denseGrad2, denseGrad1)
+          (grad1, loss1 + loss2)
```
I think `grad1` here will need to be `denseGrad1`?

AnthonyTruchet replied:
done, thanks for spotting.

SparkQA commented:
**[Test build #69469 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69469/consoleFull)** for PR 16037 at commit [`ec183a2`](https://github.com/apache/spark/commit/ec183a20166ec12c11b3c637b8722ef9cdb8bcc4).

SparkQA commented:
**[Test build #69470 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69470/consoleFull)** for PR 16037 at commit [`9b30c5c`](https://github.com/apache/spark/commit/9b30c5c8a01f7d3408eae9c27e2e3b52b216a0b2).

SparkQA commented:
**[Test build #69469 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69469/consoleFull)** for PR 16037 at commit [`ec183a2`](https://github.com/apache/spark/commit/ec183a20166ec12c11b3c637b8722ef9cdb8bcc4).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69469/
Test FAILed.

SparkQA commented:
**[Test build #69470 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69470/consoleFull)** for PR 16037 at commit [`9b30c5c`](https://github.com/apache/spark/commit/9b30c5c8a01f7d3408eae9c27e2e3b52b216a0b2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69470/
Test PASSed.

AnthonyTruchet commented:
@MLnick there it is all tests passing :-)

MLnick commented:
By the way this same issue may also impact the `ml` optimizers that use L-BFGS. We should check the various gradient aggregators for `LogisticRegression`, `LinearRegression`, `MLP` etc. cc @sethah

MLnick commented:
I'm sure this will be net positive, and _shouldn't_ cause any regression. Still, we must be certain. @AnthonyTruchet can you provide for posterity the detailed test results for the vector sizes you mentioned? And perhaps also some results for some smaller sizes (since I imagine the benefit of this change for that scenario is quite small and we should just check there is no unexpected overhead or regression we've somehow missed from the `toDense` calls though I can't see that happening).

AnthonyTruchet commented:
L-BFGS is the only optimizer I've used so far. I'm not sure how much time I can free to take care of the other ones, but I'll try :-)

Regarding the bench, I'll check if we have archived the results, otherwise I'll relaunch it next week

srowen commented:
Yes I'm pretty OK with merging this. If you can dig up any results, that's all the better. Will check in with you next week.

sethah commented on mllib/src/main/scala/org/apache/spark/mllib/optimization/LBFGS.scala:
```diff
@@ -241,16 +241,27 @@ object LBFGS extends Logging {
+       }
+
+      val zeroSparseVector = Vectors.sparse(n, Seq())
+      val (gradientSum, lossSum) = data.treeAggregate(zeroSparseVector, 0.0)(seqOp, combOp)
```
I'm slightly in favor of keeping the parentheses for the zero values. If you don't know the signature of `treeAggregate` and that the scala compiler evidently packs these into a tuple, you may be confused about what the second argument is. Not a strong preference.

AnthonyTruchet replied:
You mean `data.treeAggregate((zeroSparseVector, 0.0))`, don't you ?

sethah replied:
Yes :)

sethah commented on mllib/src/main/scala/org/apache/spark/mllib/optimization/LBFGS.scala:
```diff
@@ -241,16 +241,27 @@ object LBFGS extends Logging {
+      // Adds two (gradient, loss) tuples
+      val combOp = (c1: (Vector, Double), c2: (Vector, Double)) =>
+        (c1, c2) match { case ((grad1, loss1), (grad2, loss2)) =>
+          val denseGrad1 = grad1.toDense
```
This is necessary because we may have empty partitions, right? Might be nice to add a test explicitly for this (it seems it failed in pyspark without it, but that was just a coincidence?) so someone doesn't remove these lines in the future.

srowen replied:
Meaning, when would the args ever not be dense? I agree, shouldn't be sparse at this stage, but doing this defensively seems fine since it's a no-op for dense.

AnthonyTruchet replied:
Yes. actually missing the handling of dense vector was the cause for the PySpark UTest failure we observed.

sethah commented on mllib/src/main/scala/org/apache/spark/mllib/optimization/LBFGS.scala:
```diff
@@ -241,16 +241,27 @@ object LBFGS extends Logging {
+            (denseGrad, loss + l)
+        }
+
+      // Adds two (gradient, loss) tuples
```
I don't think this comment is necessary. The semantics of a combine op should be well understood.

AnthonyTruchet replied:
OK

sethah commented:
@MLnick Yeah, this is likely a problem with all the ML aggregators as well. We can probably take care of it using lazy evaluation.

srowen reviewed (changes requested):
Agree with @sethat's two comments regarding the comment and the arg to treeAggregate (that generates a warning I think). Then I think it's good to go.

AnthonyTruchet commented:
@sethah Agreed, that's why I uggested to add a dedicated treeAggregate wrapper to MLlin Utils which would take care of that without fiddling with sparsity for each seqOP and comboOp. See closed  #16078 for the idea...

Commit:
Stylistic review taken into account

SparkQA commented:
**[Test build #69722 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69722/consoleFull)** for PR 16037 at commit [`3b59ab2`](https://github.com/apache/spark/commit/3b59ab283797bb8448b2b4384caa0fb12ae9fece).

SparkQA commented:
**[Test build #69722 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69722/consoleFull)** for PR 16037 at commit [`3b59ab2`](https://github.com/apache/spark/commit/3b59ab283797bb8448b2b4384caa0fb12ae9fece).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69722/
Test PASSed.

sethah commented on mllib/src/main/scala/org/apache/spark/mllib/optimization/LBFGS.scala:
```diff
@@ -241,16 +241,24 @@ object LBFGS extends Logging {
+
+      val combOp = (c1: (Vector, Double), c2: (Vector, Double)) =>
+        (c1, c2) match { case ((grad1, loss1), (grad2, loss2)) =>
+          val denseGrad1 = grad1.toDense
```
I am still pretty strongly in favor of adding a test case explicitly for this. Just make an RDD with at least one empty partition, and be sure that LBFGS will run on it.

AnthonyTruchet replied:
I'm sorry, I have no clue how to generate a non-empty RDD with an empty partition. Can you please hint me at some entry points so that I can contribute the UTest you request ?

srowen replied:
What about just parallelizing n elements into more than n partitions? at least one partition must be empty.

srowen replied:
@AnthonyTruchet does that let you add a quick test case? then I think this can be merged.

AnthonyTruchet replied:
I'll try to work on this in the next two days. I will not be able to relaunch the actual benchmark we did weeks ago, our internal codebase has changed too much.

AnthonyTruchet replied:
I did thanks and I merged it. I was just comming back to this contrib when I saw it.

sethah committed:
add test

AnthonyTruchet committed:
Merge pull request #21 from sethah/pr-16037

[SPARK-18471] add test

SparkQA commented:
**[Test build #70081 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70081/consoleFull)** for PR 16037 at commit [`18fcbba`](https://github.com/apache/spark/commit/18fcbba81168c741497c57ede2554ed0c8e48d2c).

SparkQA commented:
**[Test build #70081 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70081/consoleFull)** for PR 16037 at commit [`18fcbba`](https://github.com/apache/spark/commit/18fcbba81168c741497c57ede2554ed0c8e48d2c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70081/
Test PASSed.

sethah commented:
LGTM

srowen commented:
Merged to master

AnthonyTruchet commented:
Thanks for keeping up with this merge request, I've learned a lot wrt the contribution process and good practice, and next contrib will hopefully be much more straightforward. Thanks to the Spark commiter team for this great piece of software :-)

dbtsai commented:
Sorry for late review. Just come back to US. LGTM too! Thanks.
